packetbeat/protos/tls/parse.go

```diff
@@ -0,0 +1,506 @@
+	clientHello                      = 1
+	serverHello                      = 2
+	certificate                      = 11
+	certificateRequest               = 13
+)
+
+type tlsVersion struct {
+	major, minor uint8
+}
+
+type recordHeader struct {
+	recordType uint8
+	version    tlsVersion
+	length     uint16
+}
+
+type handshakeHeader struct {
+	handshakeType handshakeType
+	length        int
+}
+
+type commonHelloFields struct {
+	version    tlsVersion
+	timestamp  uint32
+	sessionID  string
+	extensions common.MapStr
+}
+
+type clientHelloMessage struct {
+	commonHelloFields
+	cipherSuites []cipherSuite
+	compMethods  []compressionMethod
+}
+
+type serverHelloMessage struct {
+	commonHelloFields
+	cipher     cipherSuite
+	compMethod compressionMethod
+}
+
+func (header *recordHeader) read(buf *streambuf.Buffer) error {
+	var err error
+	if header.recordType, err = buf.ReadNetUint8At(0); err != nil {
+		return err
+	}
+	if header.version.major, err = buf.ReadNetUint8At(1); err != nil {
+		return err
+	}
+	if header.version.minor, err = buf.ReadNetUint8At(2); err != nil {
+		return err
+	}
+	if header.length, err = buf.ReadNetUint16At(3); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (header *handshakeHeader) read(buf *streambuf.Buffer) error {
+	var err error
+	var len8, typ uint8
+	var len16 uint16
+	if typ, err = buf.ReadNetUint8At(0); err != nil {
+		return err
+	}
+	if len8, err = buf.ReadNetUint8At(1); err != nil {
+		return err
+	}
+	if len16, err = buf.ReadNetUint16At(2); err != nil {
+		return err
+	}
+	header.handshakeType = handshakeType(typ)
+	header.length = int(len16) | (int(len8) << 16)
+	return nil
+}
+
+func (header *recordHeader) String() string {
+	return fmt.Sprintf("recordHeader type[%v] version[%v] length[%d]",
+		header.recordType, header.version, header.length)
+}
+
+func (header *recordHeader) isValid() bool {
+	return header.version.major == 3 && header.length <= maxTLSRecordLength
+}
+
+func (s commonHelloFields) toMap() common.MapStr {
+	m := common.MapStr{
+		"version":   fmt.Sprintf("%d.%d", s.version.major, s.version.minor),
+		"timestamp": s.timestamp,
+	}
+	if len(s.sessionID) != 0 {
+		m["session_id"] = s.sessionID
+	}
+	return m
+}
+
+func (s clientHelloMessage) toMap() common.MapStr {
+	m := s.commonHelloFields.toMap()
+	ciphers := make([]string, len(s.cipherSuites))
+	for idx, code := range s.cipherSuites {
+		ciphers[idx] = code.String()
+	}
+	m["ciphers"] = ciphers
+	comp := make([]string, len(s.compMethods))
+	for idx, code := range s.compMethods {
+		comp[idx] = code.String()
+	}
+	m["compression_methods"] = comp
+	m["extensions"] = s.extensions
+	return m
+}
+
+func (s serverHelloMessage) toMap() common.MapStr {
+	m := s.commonHelloFields.toMap()
+	m["cipher"] = s.cipher.String()
+	m["compression_method"] = s.compMethod.String()
+	m["extensions"] = s.extensions
+	return m
+}
+
+func (parser *parser) parse(buf *streambuf.Buffer, conn *tlsConnectionData, plugin *tlsPlugin) (bool, bool) {
```

Comment:
Try not to introduce circular dependencies in any way. At optimium the parser must be testable without requiring an instance of `tlsPlugin`. This simplifies potential unit and/or fuzzy testing of the parser only.

The TCP code generate creates different types with different responsibilities on purpose. One type per task like: manage connections, parse, correlate messages, create events, publish events